### PR TITLE
Checking events and processes using schedule and listeners

### LINKED ISSUE #28: Model transition probabilities from erlang distribution
Stochastic scheduling based on the erlang distribution.

Note: make sure schedules are cleared when individuals die

You can look at malaria model for reference

### PR DESCRIPTION
**Pull request by ceeb950**
Bringing together OJ and Giovanni's work in this area. 2 tests are failing so will need some help with them,

### CODE REVIEW

**Comment by ceeb950** [comment]
@richfitz @giovannic @OJWatson  - I have created a pull request but note 2 tests are failing so I will need help with these.

**Comment by ceeb950** [comment]
@giovannic @richfitz @OJWatson - ready for review -

File: README.Rmd [hunk 1]
```diff
@@ -120,10 +120,15 @@ This is an example showing how states are checked
   expect_equal(states$D, D)
 ```
 
+## Example
+
+```{r, echo = FALSE}
+options(warn = - 1)
```
Comment (richfitz) [inline README.Rmd]: Can you please not disable warnings like this - the code should not issue warnings, especially when used as intended.

File: man/r_erlang.Rd [hunk 2]
```diff
@@ -0,0 +1,19 @@
+% Generated by roxygen2: do not edit by hand
```
Comment (richfitz) [inline man/r_erlang.Rd:1]: Non-exported functions should not normally have .Rd files generated

File: tests/testthat/test-events.R [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+  events$iox_get_live$add_listeners(api, c(2, 4))
+  events$iox_get_die$add_listeners(api, c(2, 4))
+  events$imv_not_get_live $add_listeners(api, c(2, 4))
+  events$imv_not_get_die $add_listeners(api, c(2, 4))
+  events$iox_not_get_live$add_listeners(api, c(2, 4))
+  events$iox_not_get_die$add_listeners(api, c(2, 4))
+  events$stepdown$add_listeners(api, c(2, 4))
+  events$recovery$add_listeners(api, c(2, 4))
+  events$death$add_listeners(api, c(2, 4))
+
```
Comment (richfitz) [inline tests/testthat/test-events.R:62]: This test block appears to lack any expectations? Or am I missing some that are implicit?

File: tests/testthat/test-parameters.R [hunk 4]
```diff
@@ -1,21 +0,0 @@
-test_that("test get_parameters returns the correct values from SQUIRE", {
```
Comment (richfitz) [inline tests/testthat/test-parameters.R]: This file appears to have been deleted in error.

File: R/events.R [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+    # Vaccination events
+
+  )
+
+  events
+}
+
+#' @title Update the state of an individual as infection events occur
+#' @description Randomly moves individuals towards the later stages of disease
```
Comment (OJWatson) [inline R/events.R]:
```suggestion
#' @description Moves individuals towards the later stages of disease
```

File: R/events.R [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+#' and updates their ineffectivity
```
Comment (OJWatson) [inline R/events.R]:
```suggestion

```

File: R/events.R [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+  states,
+  events,
+  variables,
+  parameters) {
+  function(api, to_move) {
+    disc_ages <- api$get_variable(human, variables$discrete_age, to_move)
+    prob_hosp <- parameters$prob_hosp[as.integer(disc_ages)]
+    hosp <- bernoulli_multi_p(length(to_move), prob_hosp)
+
+    if(sum(hosp) > 0) {
```
Comment (richfitz) [inline R/events.R:88]: Neither of these two conditions are run in tests - can you please create a test for this?

File: R/processes.R [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+   variables
+) {
+   function(api) {
+      parameters <- api$get_parameters()
+      exposed <- api$get_state(individuals$human, states$E)
+      age <- api$get_variable(individuals$human, variables$discrete_age, exposed)
+      prob_hosp <- parameters$prob_hosp[as.integer(age)]
+      hosp <- bernoulli_multi_p(length(exposed), prob_hosp)
+
+      if(sum(hosp) > 0) {
```
Comment (richfitz) [inline R/processes.R:21]: Neither of these two bits of logic are triggered in tests, can you please add a test that covers this?

File: R/utils.R [hunk 9]
```diff
@@ -1,3 +1,42 @@
...
+#'
+#' @details Random draws from exponential distribution
+#' @param size Number of draws
+#' @param mu Mean duration
+#' @export
+#' @importFrom stats rexp
+r_exp <- function(size, mu) { rexp(size, rate = 1/mu) }
+
+#' @noRd
+remove_non_numerics <- function(l) {
```
Comment (richfitz) [inline R/utils.R:26]: This function can be removed as is not referenced from anywhere

File: R/utils.R [hunk 10]
```diff
@@ -1,3 +1,42 @@
...
+  classes <- unlist(lapply(l, class))
+  return(l[classes %in% c("numeric", "integer", "array")])
+}
+
+#' @noRd
+is_integer_like <- function(x) {
```
Comment (richfitz) [inline R/utils.R]: Unused, can be removed

File: R/utils.R [hunk 11]
```diff
@@ -1,3 +1,42 @@
...
+  abs(x - round(x)) < sqrt(.Machine$double.eps)
+}
+
+#' @noRd
+isEmpty <- function(x) {
```
Comment (richfitz) [inline R/utils.R]: Unused, can be removed

File: R/utils.R [hunk 12]
```diff
@@ -1,3 +1,42 @@
...
+r_erlang <- function(size, mu) { rgamma(size, shape = 2, rate = 2/mu) }
+
+#' Exponential waiting time distribution
+#'
+#' @details Random draws from exponential distribution
+#' @param size Number of draws
+#' @param mu Mean duration
+#' @export
+#' @importFrom stats rexp
+r_exp <- function(size, mu) { rexp(size, rate = 1/mu) }
```
Comment (richfitz) [inline R/utils.R]: Please format the above functions; spaces around the `/`, newline after `{` and `}` on its own line (not sure why codefactor is not picking this up), braces around the single line functions at the top of the file

File: NAMESPACE [hunk 13]
```diff
@@ -1,4 +1,8 @@
 # Generated by roxygen2: do not edit by hand
 
-export(create_states)
```
Comment (richfitz) [inline NAMESPACE:3]: create_states is still referenced in the README, but has been removed from the package

File: R/events.R [hunk 14]
```diff
@@ -0,0 +1,108 @@
+#' @ Create events
+#'
+#' @return events
+create_events <- function() {
+
+  events = list(
```
Comment (richfitz) [inline R/events.R]: Please use <- for assignment.

It looks like codefactor has stopped - it looks like this is possibly because of an error in the .lintr file:

```
.lintr unsupported syntax: snake_case_linter = NULL.
```

but that looks fine to me. If you run `lintr::lint_dir()` this should reveal the issues and I can investigate why it's not working after this is merged

File: R/processes.R [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+   )
+
+   # Case Infection events
+   events$severe_infection$add_listener(
+      create_progression_listener(
+         event = events$hospitilisation,
+         duration = parameters$dur_ICase
+      )
+   )
+
```
Comment (OJWatson) [inline R/processes.R]: Where is the hospitalisation listener? Has this just been removed to be added later for #9? Seems strange to do that but okay to ensure one feature per PR. 

Also I know you were debugging the hospitalisation flow yesterday - was there a specific bug you were working through? Happy to help today

File: tests/testthat/test-events.R [hunk 16]
```diff
@@ -0,0 +1,242 @@
+test_that("test all listeners", {
+
+  parameters <- mockery::mock()
+  states <-  list(E = mockery::mock(),
```
Comment (OJWatson) [inline tests/testthat/test-events.R]: Seems peculiar to have the old states here. Would remove to prevent confusion

File: NAMESPACE [hunk 17]
```diff
@@ -1,4 +1,8 @@
...
 export(get_parameters)
```
Comment (OJWatson) [inline NAMESPACE:3]: You need to rebuild as have just seen you have now deleted the old `get_parameters` function but tests will fail now as this is still in NAMESPACE

**Review by OJWatson — CHANGES_REQUESTED** [review]
Looks good Carol. Science wise looks fine and be good to get this merged in after the other reviews

**Review by richfitz — CHANGES_REQUESTED** [review]
I can't comment on the biology at all, but see that OJ has done so. Major things are - it would be good to get tests on the processes here (see comments inline) and there seems to have been some accidental deletions due to the merge. Oddly codefactor is not working - it's registering an error on its status page. I can investigate on another branch later.

File: R/events.R (hunk 7 above)
Comment (ceeb950) [reply R/events.R:88]: yes - currently doing this

**Comment by ceeb950** [comment]
@giovannic please review - there is still the mismatch with sum(pop$n) and length of continuous age.

File: tests/testthat/test-utils.R [hunk 18]
```diff
@@ -7,6 +7,22 @@ test_that("null-or-value works", {
...
+  ret <- r_exp(1, 0.1)
+  expect_true(ret >= 0.0)
+  expect_true(ret <= 1.0)
+
+})
+
+test_that("bernoulli_multi_p works", {
+
+  ret <- bernoulli_multi_p(1, 0.1)
+  expect_true(ret >= 0.0)
```
Comment (richfitz) [inline tests/testthat/test-utils.R]: this pair of tests is true for any number (which the logical return value here of either TRUE or FALSE is converted into).

So we might have:

```
expect_type(ret, "logical")
expect_equal(length(ret), 1)
```

if you want to test something about your existing `ret` but perhaps the corner cases:

```
expect_true(all(bernoulli_multi_p(10000, 1)))
expect_false(any(bernoulli_multi_p(10000, 0)))
```

and statistical expectations

```
testthat::try_again(5,
  expect_equal(mean(bernoulli_multi_p(1e6, 0.5)), 0.5, tolerance = 1e-3))
```

File: tests/testthat/test-variables.R [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+
+})
+
+test_that("test create_discrete_age_variable", {
+
+  pop <- squire::get_population(iso3c = "ATG", simple_SEIR = FALSE)
+  ages <- create_continuous_age_variable(pop = pop, max_age = 100)
+  disc_ages <- create_discrete_age_variable(ages, pop)
+
+  expect_true(all(as.numeric(table(disc_ages)) == pop$n))
```
Comment (richfitz) [inline tests/testthat/test-variables.R]:
```suggestion
  expect_equal(as.numeric(table(disc_ages), pop$n)
```

File: R/variables.R [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+#' @param pop population list
+#' @param max_age maximum age to be drawn
+#'
+#' @return continuous age variable
+#' @importFrom stats dexp
+create_continuous_age_variable <- function(pop, max_age = 100) {
+
+  # get out country median ages
+  iso3c <- pop$iso3c[1]
+  iso3c_ages <- hypatia::iso3c_ages
```
Comment (richfitz) [inline R/variables.R]: This line should be removed

File: R/variables.R [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+  r <- list()
+  for (i in seq_along(age_bins)[-1]) {
+    r[[i - 1]] <- seq(age_bins[i - 1], age_bins[i] - 1, 1)
+  }
+  r[[length(r) + 1]] <- seq(max(age_bins), max_age, 1)
+
+  # now sample from these
+  ages <- list()
+  for (i in seq_len(length(pop$age_group))) {
+    ages[[i]] <- sample(r[[i]], pop$n[i], replace = TRUE,
```
Comment (richfitz) [inline R/variables.R:28]: The reason why your tests are failing is that `pop$n` is not an integer here, but is a floating point number that *looks* like an integer and has suffered some rounding error at some point.

```
> floor(pop$n) - round(pop$n)
 [1]  0  0  0  0  0  0  0  0  0  0 -1  0  0  0  0  0  0
```

`sample` is doing an integer cast on the number which is where your -1 is coming from in the test (which shows the test is working!).

The solution is to round the numbers in your data processing script, I think

File: tests/testthat/test-events.R [hunk 22]
```diff
@@ -0,0 +1,236 @@
+test_that("create_event_based_processes assigns a listener to each event", {
+
+  pop <- get_population("Antigua and Barbuda")
+  contact_matrix_set = squire::get_mixing_matrix(iso3c = "ATG")
```
Comment (richfitz) [inline tests/testthat/test-events.R]: `<-` for assignment please

File: tests/testthat/test-events.R [hunk 23]
```diff
@@ -0,0 +1,236 @@
...
+  parameters <- get_parameters(pop, 1, 100, contact_matrix_set)
```
Comment (richfitz) [inline tests/testthat/test-events.R]: I suspect you're on it already but this is failing for me:

```
Error in parse_country_severity(country = country, prob_hosp = prob_hosp,  : 
  Country not found
In addition: Warning message:
In if (!country %in% unique(squire::population$country)) { :
  the condition has length > 1 and only the first element will be used
> traceback()
4: stop("Country not found")
3: parse_country_severity(country = country, prob_hosp = prob_hosp, 
       prob_severe = prob_severe, prob_non_severe_death_treatment = prob_non_severe_death_treatment, 
       prob_severe_death_treatment = prob_severe_death_treatment, 
       prob_non_severe_death_no_treatment = prob_non_severe_death_no_treatment, 
       prob_severe_death_no_treatment = prob_severe_death_no_treatment, 
       walker_params = walker_params)
2: squire::parameters_explicit_SEEIR(population = pop$n, dt = dt, 
       time_period = time_period, ...) at parameters.R#11
1: get_parameters(pop, 1, 100, contact_matrix_set)
```

The error (and warning) seem to come from squire though, but it looks like that is because of an error in how get_parameters is calling the squire function?

**Review by richfitz — CHANGES_REQUESTED** [review]
A pointer on the population mismatch

File: R/variables.R (hunk 21 above)
Comment (ceeb950) [reply R/variables.R:28]: ok, was tired last night when committing so couldn't see the wood for the trees. Will look now

Comment (OJWatson) [reply R/variables.R:28]: This does stem from `squire` not rounding model estimates of the demography. This was fixed in squire `v0.5.5` (https://github.com/mrc-ide/squire/blob/master/tests/testthat/test-population.R#L108-L113) so change the dependency in DESCRIPTION as well.

**Comment by ceeb950** [comment]
@giovanni - please review. The coverage is 100% for the tests  but events.R is covered with an almost identical test twice, once for hosp TRUE then for FALSE. All tests now pass.

**Comment by ceeb950** [comment]
@richfitz Rich - review ready - I have asked Giovanni to look at two tests covering two forks of the same function. I could not seem to simplify it so have 2 tests instead of one with two parts.

File: NAMESPACE [hunk 24]
```diff
@@ -1,5 +1,10 @@
...
+export(get_population)
+export(r_exp)
```
Comment (giovannic) [inline NAMESPACE]: We don't need to export r_exp

File: R/events.R [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+    iox_get_die = individual::Event$new('iox_get_die'),
+    imv_not_get_live = individual::Event$new('imv_not_get_live'),
+    imv_not_get_die = individual::Event$new('imv_not_get_die'),
+    iox_not_get_live = individual::Event$new('iox_not_get_live'),
+    iox_not_get_die = individual::Event$new('iox_not_get_die'),
+    stepdown = individual::Event$new('stepdown'),
+    recovery = individual::Event$new('recovery'),
+    death = individual::Event$new('deaths')
+
+    # Vaccination events
```
Comment (giovannic) [inline R/events.R]: Can we remove this comment?

File: R/events.R [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+}
+
+#' @title Modelling the progression of the human disease
+#' @description schedules follow up infection events
+#'
+#' @param event the event to schedule
+#' @param duration the average time spent in this state
+#' @param shift number of days to increase scheduled event by
+#' @param func function to use for drawing progression.
+#'   Default = \code{\link{r_erlang}}
```
Comment (giovannic) [inline R/events.R]:
```suggestion
#'   Default = [r_erlang]
```

File: R/individuals.R [hunk 27]
```diff
@@ -1,10 +1,26 @@
...
+  parameters) {
+
+  human <- individual::Individual$new(
+    "human",
+    states = states,
+    variables = variables,
+    events = events
+  )
+
+  list(human = human)
```
Comment (giovannic) [inline R/individuals.R]: [nit] We can just return human no?

File: R/parameters.R [hunk 28]
```diff
@@ -1,28 +1,39 @@
 #' @title Get parameters from SQUIRE model
 #'
 #' @param pop population. See [squire::get_population]
-#' @param dt the timestep (days) for the simulation
+#' @param R0 R0
 #' @param time_period the number timesteps to run the simulation for
-#' @param ... Other parameters for [squire::parameters_explicit_SEEIR] 
+#' @param tt_contact_matrix tt_contact_matrix
```
Comment (giovannic) [inline R/parameters.R]: tt_contact_matrix and contact_matrix_set are dealt with in paramters_explicit_SEEIR. It feels redundant to deal with them here too.

Can we leave it how it was?

File: DESCRIPTION [hunk 29]
```diff
@@ -20,4 +20,9 @@ Suggests:
     mockery
 Imports:
     individual,
-    squire
+    squire,
```
Comment (richfitz) [inline DESCRIPTION]: Given this comment https://github.com/mrc-ide/hypatia/pull/32#discussion_r530936805 can we please increase the squire version number as it looks like that will fix the underlying issue.

```suggestion
    squire (>= 0.5.5),
```

and the Remotes to

```
    mrc-ide/squire@v0.5.5
```

We've set up squire now to automatically tag on release

File: tests/testthat/test-events.R [hunk 30]
```diff
@@ -0,0 +1,293 @@
...
+    parameters
+  )
+
+  for (event in events) {
+    expect_gt(length(event$listeners), 0)
+  }
+
+})
+
+test_that("test that listeners do not call empty targets", {
```
Comment (giovannic) [inline tests/testthat/test-events.R]: This tests the opposite. We call the listeners with empty targets and queue_state_update is called

File: tests/testthat/test-model-integration.R [hunk 31]
```diff
@@ -1,5 +1,5 @@
-test_that("run_simulation can parameterise and run an Afghan model for 10 days", {
```
Comment (giovannic) [inline tests/testthat/test-model-integration.R:1]: This still works on the Afghanistan data

File: tests/testthat/test-model-integration.R [hunk 32]
```diff
@@ -1,5 +1,5 @@
...
-  expected_columns <- c('timestep')
```
Comment (giovannic) [inline tests/testthat/test-model-integration.R:2]: [nit] this variable name is explanatory, why are we removing it?

File: R/variables.R [hunk 33]
```diff
@@ -0,0 +1,86 @@
...
+#' @description Create all individual variables for humans
+#'
+#' @param pop population list
+#' @param max_age maximum age - default 100
+#'
+#' @return named list of individual::Variable
+create_variables <- function(pop, max_age = 100) {
+
+  ret <- create_age_variables(pop, max_age = 100)
+  list(age = ret$age, discrete_age = ret$discrete_age)
```
Comment (giovannic) [inline R/variables.R]: "age" is unused

File: tests/testthat/test-events.R [hunk 34]
```diff
@@ -0,0 +1,293 @@
...
+
+  r_erlang_mock <- mockery::mock(c(TRUE, TRUE, TRUE, TRUE))
+  mockery::stub(ret, 'r_erlang', mockery::mock(c(TRUE, TRUE, TRUE, TRUE)))
+
+  mockery::stub(ret, 'bernoulli_multi_p', mockery::mock(TRUE))
+
+  mockery::stub(
+    ret,
+    "hosp",
+    mockery::mock()(
```
Comment (giovannic) [inline tests/testthat/test-events.R]: I'm not sure what we're trying to do here but it doesn't look like the correct use of `mockery::mock`. I'll ask offline

**Review by giovannic — CHANGES_REQUESTED** [review]
We're closer.

 * Two of the tests are still confusing
 * There's a bit of redundant code
 * I've used [nit] to show where I might be pedantic, take those suggestions as you want
 * I've recently learned about `@noRd` and how it's useful for suppressing documentation on internal functions. We should try and implement that here.

Let me know when you have time to walk you through the comments, I can give you extra context on a call

File: R/parameters.R (hunk 28 above)
Comment (ceeb950) [reply R/parameters.R]: it didn't work that's why I added them  - talk offline

File: R/variables.R (hunk 33 above)
Comment (ceeb950) [reply R/variables.R]: I understand it will be used. talk offline

**Comment by ceeb950** [comment]
@richfitz - Rich, all tests passing, 100% coverage and reviewed with Giovanni -https://github.com/mrc-ide/hypatia/pull/32

File: tests/testthat/test-utils.R [hunk 35]
```diff
@@ -7,6 +7,29 @@ test_that("null-or-value works", {
...
+  expect_true(ret <= 1.0)
```
Comment (richfitz) [inline tests/testthat/test-utils.R]: these two lines should still come out as they are unfalsifiable

File: R/variables.R [hunk 36]
```diff
@@ -0,0 +1,86 @@
...
+  r <- list()
+  for (i in seq_along(age_bins)[-1]) {
+    r[[i - 1]] <- seq(age_bins[i - 1], age_bins[i] - 1, 1)
+  }
+  r[[length(r) + 1]] <- seq(max(age_bins), max_age, 1)
+
+  # now sample from these
+  ages <- list()
+  for (i in seq_len(length(pop$age_group))) {
+    ages[[i]] <- sample(r[[i]], round(pop$n[i]), replace = TRUE,
```
Comment (richfitz) [inline R/variables.R]: With the updated squire, this round should not be needed any more

**Review by richfitz — CHANGES_REQUESTED** [review]
One tiny thing that will fix the population problem and keep the implementation a little simpler

**Review by richfitz — APPROVED** [review]
I just tweaked my requested changes, hope you don't mind

**Review by OJWatson — APPROVED** [review]

**Comment by ceeb950** [comment]
@giovannic @richfitz  - changes made as requested - please finalise so I can merge with main  https://github.com/mrc-ide/hypatia/pull/32

**Review by giovannic — APPROVED** [review]

**Comment by ceeb950** [comment]
close #32

**Comment by ceeb950** [comment]
close #32
